federated snakemake workflow

vschnei opened the pull request:
This branch adds a snakemake workflow  which processes images at specific (geo proximity) TES instances and saves the results in a S3 storage.

Commit:
add federated workflow

add tutorian on workflow submission via WESkit

uniqueg commented on wf-federated/config.yaml:
```diff
@@ -0,0 +1,5 @@
+urls:
+  finlad: "https://csc.fi/o/csc-theme/images/csc-logo-teksti-fi.png"
```
`finland`, not `finlad`

vschnei replied:
right!

uniqueg commented on wf-federated/config.yaml:
```diff
@@ -0,0 +1,5 @@
+urls:
+  finlad: "https://csc.fi/o/csc-theme/images/csc-logo-teksti-fi.png"
+  greece: "https://gsri.gov.gr/wp-content/uploads/2022/05/Athena_Horizontal-Ext_EN-page-001-3.jpg"
+  czech: "https://www.mzv.cz/public/a8/13/d5/661666_637430_mzv_en.png"
```
I think we should perhaps rather write `czech_republic`. People sometimes feel ways about things like these, and for the other countries, we use the proper country name and not the adjective.

vschnei replied:
done.

uniqueg commented on wf-federated/config.yaml:
```diff
@@ -0,0 +1,5 @@
+  finlad: "https://csc.fi/o/csc-theme/images/csc-logo-teksti-fi.png"
+  greece: "https://gsri.gov.gr/wp-content/uploads/2022/05/Athena_Horizontal-Ext_EN-page-001-3.jpg"
+  czech: "https://www.mzv.cz/public/a8/13/d5/661666_637430_mzv_en.png"
+  united_kindom: "http://britishfamily.co.uk/wp-content/uploads/2015/02/MADE_IN_BRITAIN_web_300x300.png"
```
`kingdom`, not `kindom`

vschnei replied:
right!

uniqueg commented on wf-federated/Snakefile:
```diff
@@ -0,0 +1,44 @@
+                      endpoint_url="https://s3-elixir.cloud.e-infra.cz")
+
+# adjust to your own path
+outfile_s3 = "BUCKET/PATH/demo_outfile_result.txt"
+remote_sizes_file_s3 = "BUCKET/PATH/demo_sizes_file.txt"
+remote_out_file_s3 = "BUCKET/PATH/demo_logo.png"
```
Any way we can use an env var for `BUCKET/PATH`? Again, it would be great if we could separate the config from the descriptor file.

vschnei replied:
done

uniqueg commented on wf-federated/Snakefile:
```diff
@@ -0,0 +1,44 @@
+
+S3 = S3RemoteProvider(access_key_id=os.environ["ACCESS_KEY_ID"], 
+                      secret_access_key=os.environ["SECRET_ACCESS_KEY"], 
+                      endpoint_url="https://s3-elixir.cloud.e-infra.cz")
```
In order to separate configuration from the descriptor file entirely, maybe we could set a config var for the URL as well?

vschnei replied:
done. Downstream paths adjusted as well.

uniqueg commented on wf-federated/Snakefile:
```diff
@@ -0,0 +1,44 @@
+configfile: "/data/config.yaml"
```
So this is what Snakemake will fall back to if the config file location is not explicitly passed, correct? But using an absolute path here will almost never work. Shouldn't we use a default that points to the `config.yaml` file we ship with this repo, i.e., `configfile: "./config.yaml` in this case?

vschnei replied:
This is a remnant from the tutorial I gave. There I put the config.yaml in /data. 
adjusted

uniqueg commented on wf-federated/Snakefile:
```diff
@@ -0,0 +1,44 @@
+
+from snakemake.remote.S3 import RemoteProvider as S3RemoteProvider
+
+# any other remote storage is possible as well e.g. FTP
+# Remote storage has to be suported by Snakemake.
```
Maybe simplify into one line, e.g.:
"For other supported storage providers (e.g., FTP), see:"

vschnei replied:
Sure.

uniqueg commented on wf-hashsplitter/Snakefile:
I'm not quite sure whether we still need this file at all. I think that perhaps it'd be good if the repo we wish to include in the paper is as small as possible. The more we show, the more questions may come and the more users might be confused.

Rather, we can use [this repo](https://github.com/elixir-cloud-aai/elixir-cloud-demos) to host multiple demos.

vschnei replied:
We could do that. But then we would need to move the Snakefiles from the other demos (hashsplitter) as well.

uniqueg replied:
Was the demo ever presented? If so, we could include it as a separate demo in the table available in the other repo's `README.md`. In that case, it would be good if we had a commit representing the state that was demonstrated - so that we can recreate that in the other repo.

Then, in _this_ repo, we could minimize everything to just what's needed for the demo described in the paper (and maybe dump that on Zenodo so that we get long term storage and a DOI for that particular state).

Commit:
snakefile & config  adjustment

uniqueg committed:
Update Snakefile

- run `snakefmt`
- reorder imports so as to be compliant with PEP8
- remove outdated/trivial comments
- wrap long shell commands

uniqueg commented on demo.ipynb:
Actually, I think we also don't need the WESkit stuff here, as we are not describing it in _this_ paper. We will of course include WES (and WESkit, if you like) if we describe the ELIXIR Cloud in a follow up paper. So again, I think this new addition ("3.2 Execute the federated workflow (S3 storage) using WESkit)" should go in the other repo, as a new demo.

uniqueg committed:
minimal version (#4)

Signed-off-by: Valentin Schneider-Lunitz <valentin.schneider-lunitz@charite.de>
Co-authored-by: Valentin Schneider-Lunitz <valentin.schneider-lunitz@charite.de>
